src/transport/UnauthenticatedSessionTable.h

```diff
@@ -0,0 +1,204 @@
+#include <lib/support/ReferenceCountedHandle.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <system/TimeSource.h>
+#include <transport/raw/PeerAddress.h>
+
+namespace chip {
+namespace Transport {
+
+class UnauthenticatedSession;
+using UnauthenticatedSessionHandle = ReferenceCountedHandle<UnauthenticatedSession>;
+
+class UnauthenticatedSessionDeleter
+{
+public:
+    // This is a no-op because life-cycle of UnauthenticatedSessionTable is rotated by LRU
+    static void Release(UnauthenticatedSession * entry) {}
+};
+
+/**
+ * @brief
+ *   An UnauthenticatedSession stores the binding of TransportAddress, and message counters.
+ *
+ *   The entries are rotated using LRU, but entry can be hold by using UnauthenticatedSessionHandle, which increase the reference
+ * count by 1. If the reference count is not 0, the entry won't be pruned.
+ */
+class UnauthenticatedSession : public ReferenceCounted<UnauthenticatedSession, UnauthenticatedSessionDeleter>
+{
+public:
+    UnauthenticatedSession(const PeerAddress & address) : mPeerAddress(address) {}
+
+    UnauthenticatedSession(const UnauthenticatedSession &) = delete;
+    UnauthenticatedSession & operator=(const UnauthenticatedSession &) = delete;
+    UnauthenticatedSession(UnauthenticatedSession &&)                  = delete;
+    UnauthenticatedSession & operator=(UnauthenticatedSession &&) = delete;
+
+    uint64_t GetLastActivityTimeMs() const { return mLastActivityTimeMs; }
+    void SetLastActivityTimeMs(uint64_t value) { mLastActivityTimeMs = value; }
+
+    const PeerAddress & GetPeerAddress() const { return mPeerAddress; }
+
+    MessageCounter & GetLocalMessageCounter() { return mLocalMessageCounter; }
+    PeerMessageCounter & GetPeerMessageCounter() { return mPeerMessageCounter; }
+
+private:
+    uint64_t mLastActivityTimeMs = 0;
+
+    const PeerAddress mPeerAddress;
+    GlobalUnencryptedMessageCounter mLocalMessageCounter;
+    PeerMessageCounter mPeerMessageCounter;
+};
+
+template <size_t kMaxConnectionCount, Time::Source kTimeSource = Time::Source::kSystem>
+class UnauthenticatedSessionTable
+{
+public:
+    /**
+     * Allocates a new session out of the internal resource pool.
+     *
+     * @returns CHIP_NO_ERROR if new session created. May fail if maximum connection count has been reached (with
+     * CHIP_ERROR_NO_MEMORY).
+     */
+    CHECK_RETURN_VALUE
+    CHIP_ERROR AllocEntry(const PeerAddress & address, UnauthenticatedSession *& entry)
+    {
+        entry = mEntries.CreateObject(address);
+        if (entry != nullptr)
+            return CHIP_NO_ERROR;
+
+        entry = FindLeastRecentUsedEntry();
+        if (entry == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+        if (currentTime - entry->GetLastActivityTimeMs() < kMinimalActivityTimeMs)
+        {
+            // Protect the entry for a short period to prevent from rotating too fast.
```

Comment:
I don't think we should throttle like this.  We are in an event loop.  If we're using the entry right now, we will be executing in the loop and blocking other activity.  If we're not, we need a better strategy to block eviction anyway.  For instance, the spec implies we can only have one PASE session open at a time, which further implies it should be impossible to evict an active PASE session.

This throttling mechanism is a blunt and imperfect mechanism to achieve that anyway.  And it appears the reference counting is the thing that would actually enforce this.  I think we should just remove the throttling (`kMinimalActivityTimeMs` etc.) altogether.

Reply:
Applied